File: 08_Principal_Component_Analysis_and_Clustering.py
## load packages
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt
import matplotlib
import numpy as np
import os
from sklearn.impute import KNNImputer
from sklearn.decomposition import PCA
from scipy.cluster.vq import vq, kmeans

## download, format, and save data
chap_8_combine_py_file = "./data/chap_8_combine_py_file.csv"
if os.path.isfile(chap_8_combine_py_file):
    combine_py = pd.read_csv(chap_8_combine_py_file)
else:
    combine_py = pd.DataFrame()
    for i in range(2000, 2023 + 1):
        url = "https://www.pro-football-reference.com/draft/" + str(i) + "-combine.htm"
        web_data = pd.read_html(url)[0]
        web_data["Season"] = i
        web_data = web_data.query('Ht != "Ht"')
        combine_py = pd.concat([combine_py, web_data])
    combine_py.reset_index(drop=True, inplace=True)

combine_py[["Ht-ft", "Ht-in"]] = combine_py["Ht"].str.split("-", expand=True)

combine_py = combine_py.astype(
    {
        "Wt": float,
        "40yd": float,
        "Vertical": float,
        "Bench": float,
        "Broad Jump": float,
        "3Cone": float,
        "Shuttle": float,
        "Ht-ft": float,
        "Ht-in": float,
    }
)
combine_py["Ht"] = combine_py["Ht-ft"] * 12.0 + combine_py["Ht-in"]
combine_py.drop(["Ht-ft", "Ht-in"], axis=1, inplace=True)

combine_py.to_csv("combine_data_py.csv", index=False)
combine_py.describe()

## several plots to look at combine data
sns.set_theme(style="whitegrid", palette="colorblind")

sns.regplot(data=combine_py, x="Ht", y="Wt")
plt.show()

sns.regplot(data=combine_py, x="Wt", y="40yd", line_kws={"color": "red"})
plt.show()

sns.regplot(data=combine_py, x="40yd", y="Vertical", line_kws={"color": "red"})
plt.show()

sns.regplot(data=combine_py, x="40yd", y="3Cone", line_kws={"color": "red"})
plt.show()

## impute missing data

combine_knn_py_file = "combine_knn_py.csv"
col_impute = ["Ht", "Wt", "40yd", "Vertical", "Bench", "Broad Jump", "3Cone", "Shuttle"]

if not os.path.isfile(combine_knn_py_file):
    combine_knn_py = combine_py.drop(col_impute, axis=1)
    imputer = KNNImputer(n_neighbors=10)
    knn_out_py = imputer.fit_transform(combine_py[col_impute])
    knn_out_py = pd.DataFrame(knn_out_py)
    knn_out_py.columns = col_impute
    combine_knn_py = pd.concat([combine_knn_py, knn_out_py], axis=1)
    combine_knn_py.to_csv(combine_knn_py_file)
else:
    combine_knn_py = pd.read_csv(combine_knn_py_file)

## PCA
pca_wt_ht = PCA(svd_solver="full")
wt_ht_py = combine_py[["Wt", "Ht"]].query("Wt.notnull() & Ht.notnull()").copy()
pca_fit_wt_ht_py = pca_wt_ht.fit_transform(wt_ht_py)

print(pca_wt_ht.explained_variance_ratio_)

plt.plot(pca_fit_wt_ht_py[:, 0], pca_fit_wt_ht_py[:, 1], "o")
plt.show()

pca_wt_ht.components_

scaled_combine_knn_py = (
    combine_knn_py[col_impute] - combine_knn_py[col_impute].mean()
) / combine_knn_py[col_impute].std()

### PCA on all data
pca = PCA(svd_solver="full")
pca_fit_py = pca.fit_transform(scaled_combine_knn_py)

rotation = pd.DataFrame(pca.components_, index=col_impute)
print(rotation)

print(pca.explained_variance_)

pca_percent_py = pca.explained_variance_ratio_.round(4) * 100
print(pca_percent_py)

pca_fit_py = pd.DataFrame(pca_fit_py)
pca_fit_py.columns = ["PC" + str(x + 1) for x in range(len(pca_fit_py.columns))]
combine_knn_py = pd.concat([combine_knn_py, pca_fit_py], axis=1)

sns.scatterplot(data=combine_knn_py, x="PC1", y="PC2")
plt.show()

sns.scatterplot(data=combine_knn_py, x="PC1", y="PC2", hue="PC3")
plt.show()


sns.scatterplot(data=combine_knn_py, x="PC1", y="PC2", hue="Pos")
plt.show()

## combine clustering data
k_means_fit_py = kmeans(combine_knn_py[["PC1", "PC2"]], 6, seed=1234)

combine_knn_py["cluster"] = vq(combine_knn_py[["PC1", "PC2"]], k_means_fit_py[0])[0]

print(
    combine_knn_py.query("cluster == 1")
    .groupby("Pos")
    .agg({"Ht": ["count", "mean"], "Wt": ["count", "mean"]})
)

combine_knn_py_cluster = combine_knn_py.groupby(["cluster", "Pos"]).agg(
    {"Ht": ["count", "mean"], "Wt": ["mean"]}
)

combine_knn_py_cluster.columns = list(map("_".join, combine_knn_py_cluster.columns))

combine_knn_py_cluster.reset_index(inplace=True)

combine_knn_py_cluster.rename(
    columns={"Ht_count": "n", "Ht_mean": "Ht", "Wt_mean": "Wt"}, inplace=True
)

combine_knn_py_cluster.cluster = combine_knn_py_cluster.cluster.astype(str)

sns.catplot(
    combine_knn_py_cluster, x="n", y="Pos", col="cluster", col_wrap=3, kind="bar"
)
plt.show()

## update figure for book
plt.figure(figsize=(6, 4.25), dpi=600)
sns.set(font_scale=1.5)
sns.catplot(
    combine_knn_py_cluster,
    x="n",
    y="Pos",
    col="cluster",
    col_wrap=3,
    kind="bar",
)
plt.savefig("fig_8_18.png", dpi=600)

combine_knn_py_cluster.groupby("cluster").agg({"Ht": ["mean"], "Wt": ["mean"]})

File: 08_Principal_Component_Analysis_and_Clustering.R
## load packages
library(tidyverse)
library(rvest)
library(htmlTable)
library(multiUS)
library(ggthemes)
library(RColorBrewer)


chap_8_combine_r_file <- "./data/chap_8_combine_r_file.csv"
if (!file.exists(chap_8_combine_r_file)) {
    combine_r <- tibble()
    for (i in seq(from = 2000, to = 2023)) {
        url <- paste0(
            "https://www.pro-football-reference.com/draft/",
            i, "-combine.htm"
        )
        web_data <-
            read_html(url) |>
            html_table()
        web_data_clean <-
            web_data[[1]] |>
            mutate(Season = i) |>
            filter(Ht != "Ht")
        combine_r <- bind_rows(combine_r, web_data_clean)
    }
    write_csv(combine_r, chap_8_combine_r_file)
} else {
    combine_r <- read_csv(chap_8_combine_r_file)
}

combine_r <-
    combine_r |>
    mutate(
        ht_ft = as.numeric(str_sub(Ht, 1, 1)),
        ht_in = str_sub(Ht, 2, 4),
        ht_in = as.numeric(str_remove(ht_in, "-")),
        Ht = ht_ft * 12 + ht_in
    ) |>
    select(-ht_ft, -ht_in)

summary(combine_r)

ggplot(combine_r, aes(x = Ht, y = Wt)) +
    geom_point() +
    theme_bw() +
    xlab("Player Height (inches)") +
    ylab("Player Weight (pounds)") +
    geom_smooth(method = "lm", formula = y ~ x)

ggplot(combine_r, aes(x = Wt, y = `40yd`)) +
    geom_point() +
    theme_bw() +
    xlab("Player Weight (pounds)") +
    ylab("Player 40-yard dash (seconds)") +
    geom_smooth(method = "lm", formula = y ~ x)

ggplot(combine_r, aes(x = `40yd`, y = Vertical)) +
    geom_point() +
    theme_bw() +
    xlab("Player 40-yard dash (seconds)") +
    ylab("Player vertical jump (inches)") +
    geom_smooth(method = "lm", formula = y ~ x)

## save figure for book
fig_8_6 <-
    ggplot(combine_r, aes(x = `40yd`, y = Vertical)) +
    geom_point() +
    theme_bw() +
    xlab("Player 40-yard dash (seconds)") +
    ylab("Player vertical jump (inches)") +
    geom_smooth(method = "lm", formula = y ~ x)
ggsave("fig_8_6.png", fig_8_6, width = 6, height = 4.25)

ggplot(combine_r, aes(x = `40yd`, y = `3Cone`)) +
    geom_point() +
    theme_bw() +
    xlab("Player 40-yard dash (seconds)") +
    ylab("Player 3 cone drill (inches)") +
    geom_smooth(method = "lm", formula = y ~ x)

## save figure for book
fig_8_9 <-
    ggplot(combine_r, aes(x = `40yd`, y = `3Cone`)) +
    geom_point() +
    theme_bw() +
    xlab("Player 40-yard dash (seconds)") +
    ylab("Player 3 cone drill (inches)") +
    geom_smooth(method = "lm", formula = y ~ x)
ggsave("fig_8_9.png", fig_8_9, width = 6, height = 4.25)


## impute missing data
combine_knn_r_file <- "combine_knn_r.csv"
if (!file.exists(combine_knn_r_file)) {
    imput_input <-
        combine_r |>
        select(Ht:Shuttle) |>
        as.data.frame()
    knn_out_r <-
        KNNimp(imput_input, k = 10, scale = TRUE, meth = "median") |>
        as_tibble()
    combine_knn_r <-
        combine_r |>
        select(Player:College, Season) |>
        bind_cols(knn_out_r)
    write_csv(x = combine_knn_r, file = combine_knn_r_file)
} else {
    combine_knn_r <- read_csv(combine_knn_r_file)
}

combine_knn_r |>
    summary()

## PCA
wt_ht_r <-
    combine_r |>
    select(Wt, Ht) |>
    filter(!is.na(Wt) & !is.na(Ht))
pca_fit_wt_ht_r <-
    prcomp(wt_ht_r)

summary(pca_fit_wt_ht_r)

pca_fit_wt_ht_r$x |>
    as_tibble() |>
    ggplot(aes(x = PC1, y = PC2)) +
    geom_point() +
    theme_bw()

print(pca_fit_wt_ht_r)

## PCA on all data
scaled_combine_knn_r <-
    scale(combine_knn_r |>
        select(Ht:Shuttle))
pca_fit_r <-
    prcomp(scaled_combine_knn_r)

print(pca_fit_r$rotation)

print(pca_fit_r$sdev^2)

pca_var_r <-
    pca_fit_r$sdev^2
pca_percent_r <-
    round(pca_var_r / sum(pca_var_r) * 100, 2)
print(pca_percent_r)

combine_knn_r <-
    combine_knn_r |>
    bind_cols(pca_fit_r$x)

ggplot(combine_knn_r, aes(x = PC1, y = PC2)) +
    geom_point() +
    theme_bw() +
    xlab(paste0("PC1 = ", pca_percent_r[1], "%")) +
    ylab(paste0("PC2 = ", pca_percent_r[2], "%"))

ggplot(combine_knn_r, aes(x = PC1, y = PC2, color = PC3)) +
    geom_point() +
    theme_bw() +
    xlab(paste0("PC1 = ", pca_percent_r[1], "%")) +
    ylab(paste0("PC2 = ", pca_percent_r[2], "%")) +
    scale_color_continuous(paste0(
        "PC3 = ",
        pca_percent_r[3], "%"
    ), low = "skyblue", high = "navyblue")


color_count <- length(unique(combine_knn_r$Pos))
get_palette <- colorRampPalette(brewer.pal(9, "Set1"))

ggplot(combine_knn_r, aes(x = PC1, y = PC2, color = Pos)) +
    geom_point(alpha = 0.75) +
    theme_bw() +
    xlab(paste0("PC1 = ", pca_percent_r[1], "%")) +
    ylab(paste0("PC2 = ", pca_percent_r[2], "%")) +
    scale_color_manual("Player position", values = get_palette(color_count))

## Combine cluster data
set.seed(123)
k_means_fit_r <-
    kmeans(
        combine_knn_r |>
            select(PC1, PC2),
        centers = 6, iter.max = 10
    )

combine_knn_r <-
    combine_knn_r |>
    mutate(cluster = k_means_fit_r$cluster)

combine_knn_r |>
    select(Pos, Ht:Shuttle, cluster) |>
    head()

combine_knn_r |>
    filter(cluster == 1) |>
    group_by(Pos) |>
    summarize(n = n(), Ht = mean(Ht), Wt = mean(Wt)) |>
    arrange(-n) |>
    print(n = Inf)

combine_knn_r_cluster <-
    combine_knn_r |>
    group_by(cluster, Pos) |>
    summarize(
        n = n(), Ht = mean(Ht),
        Wt = mean(Wt), .groups = "drop"
    )

combine_knn_r_cluster |>
    ggplot(aes(x = n, y = Pos)) +
    geom_col(position = "dodge") +
    theme_bw() +
    facet_wrap(vars(cluster)) +
    theme(strip.background = element_blank()) +
    ylab("Position") +
    xlab("Count")

## update figure for book
fig_8_19 <-
    combine_knn_r_cluster |>
    ggplot(aes(x = n, y = Pos)) +
    geom_col(position = "dodge") +
    theme_bw() +
    facet_wrap(vars(cluster)) +
    theme(
        strip.background = element_blank(),
        axis.text = element_text(size = 9)
    ) +
    ylab("Position") +
    xlab("Count")
ggsave("fig_8_19.png", fig_8_19, width = 6, height = 4.75, dpi = 600)

combine_knn_r_cluster |>
    group_by(cluster) |>
    summarize(ave_ht = mean(Ht), ave_wt = mean(Wt))
